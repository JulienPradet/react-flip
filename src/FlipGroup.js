import Flip from './Flip';

const getElementKey = defer => defer ? 'elementsDeferred' : 'elements';

const getConcernedElements = (defer, elements, deferredElements) => {
  if (defer === undefined) {
    return elements.concat(deferredElements);
  } else if (defer) {
    return deferredElements.concat(deferredElements);
  } else {
    return elements;
  }
};

class FlipGroup {
  constructor() {
    this.elements = [];
    this.elementsDeferred = [];
  }

  addElement(element, defer) {
    if (element instanceof Flip) {
      const elementKey = getElementKey(defer);
      this[elementKey] = [...this[elementKey], element];

      return () => {
        this[elementKey] = this[elementKey].filter(
          current => current !== element
        );
      };
    }
  }

  first({ deferred } = {}) {
    getConcernedElements(
      deferred,
      this.elements,
      this.elementsDeferred
    ).forEach(element => element.first());
  }

  last({ deferred } = {}) {
    getConcernedElements(
      deferred,
      this.elements,
      this.elementsDeferred
    ).forEach(element => element.last());
  }

  invert({ deferred } = {}) {
    return getConcernedElements(deferred, this.elements, this.elementsDeferred)
      .map(element => element.invert())
      .some(hasInverted => hasInverted);
  }

  play() {
    const playPromises = this.elements
      .concat(this.elementsDeferred)
      .map(element => element.play())
      .filter(promise => promise);

    if (playPromises.length === 0) {
      return;
    }
    return Promise.all(playPromises);
  }
}

export default FlipGroup;
