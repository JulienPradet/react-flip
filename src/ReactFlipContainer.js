import { Children, Component, PropTypes } from 'react';
import FlipGroup from './FlipGroup';
import Flip from './Flip';

const getDeferType = deferred =>
  deferred === undefined ? '' : deferred ? '(deferred)' : '(normal)';

const shouldAnimate = props =>
  typeof props.shouldAnimate === 'undefined' ||
  (typeof props.shouldAnimate === 'function'
    ? props.shouldAnimate(props)
    : props.shouldAnimate);

export const STATIC = 'ReactFlip_static';
export const BEFORE_ANIMATION = 'ReactFlip_before';
export const ANIMATION = 'ReactFlip_animation';

class ReactFlipContainer extends Component {
  constructor(props) {
    super();
    this.state = {
      animating: false,
      preparingAnimation: false,
      status: STATIC
    };
    this.flip = new FlipGroup({ debug: props.debug });
    this.onAnimationStartCallbacks = [];
    this.onAnimationEndCallbacks = [];
    this.registerElement = this.registerElement.bind(this);
    this.triggerAnimation = this.triggerAnimation.bind(this);
  }

  getChildContext() {
    return {
      flip: {
        status: () => this.state.status,
        registerElement: this.registerElement
      }
    };
  }

  registerElement({ element, options }) {
    const flip = new Flip({ element, options, debug: this.props.debug });

    const defer = typeof options === 'function'
      ? () => options().defer
      : options.defer;

    return this.flip.addElement(flip, defer);
  }

  componentWillReceiveProps(nextProps) {
    if (shouldAnimate(nextProps) && this.props !== nextProps) {
      if (nextProps.defer) {
        this.first({ deferred: false });
        this.setState({
          animating: false,
          preparingAnimation: true,
          status: BEFORE_ANIMATION
        });
      } else {
        this.first();
        this.setState({
          animating: true,
          preparingAnimation: false,
          status: ANIMATION
        });
      }
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (shouldAnimate(this.props)) {
      if (this.props !== prevProps) {
        if (this.props.defer) {
          // Set correct position of the undeferred elements
          this.last({ deferred: false });
          this.invert({ deferred: false });

          // Set initial position of the deferred elements
          this.first({ deferred: true });

          // The animation will now begin on next update
          this.setState({
            animating: true,
            preparingAnimation: false,
            status: ANIMATION
          });
        } else {
          this.triggerAnimation();
        }
      } else if (
        prevState.preparingAnimation !== this.state.preparingAnimation
      ) {
        this.triggerAnimation();
      }
    }
  }

  triggerAnimation() {
    this.last();
    if (this.invert()) {
      this.play();
    }
  }

  first({ deferred } = {}) {
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      const type = getDeferType(deferred);
      console.groupCollapsed(`ReactFlip: First ${type}`);
    }
    this.flip.first({ deferred });
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      console.groupEnd();
    }
  }

  last({ deferred } = {}) {
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      const type = getDeferType(deferred);
      console.groupCollapsed(`ReactFlip: Last ${type}`);
    }
    this.flip.last({ deferred });
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      console.groupEnd();
    }
  }

  invert({ deferred } = {}) {
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      const type = getDeferType(deferred);
      console.groupCollapsed(`ReactFlip: Invert ${type}`);
    }
    const result = this.flip.invert({ deferred });
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      console.groupEnd();
    }
    return result;
  }

  play() {
    if (process.env.NODE_ENV === 'development' && this.props.debug) {
      console.groupCollapsed('ReactFlip: Play');
    }
    const playPromise = this.flip.play();

    if (playPromise) {
      this.onAnimationStart()
        .then(() => playPromise)
        .then(() => {
          if (process.env.NODE_ENV === 'development' && this.props.debug) {
            console.groupEnd();
          }
        })
        .then(() => this.onAnimationEnd());
    } else {
      if (process.env.NODE_ENV === 'development' && this.props.debug) {
        console.debug('ReactFlip: Nothing to play');
        console.groupEnd();
      }
    }
  }

  onAnimationStart() {
    return Promise.resolve();
  }

  onAnimationEnd() {
    return new Promise((resolve, reject) => {
      this.setState({ animating: false, status: STATIC }, resolve);
    });
  }

  render() {
    return Children.only(this.props.children);
  }
}

ReactFlipContainer.propTypes = {
  defer: PropTypes.bool,
  children: PropTypes.node.isRequired
};

ReactFlipContainer.defaultProps = {
  defer: false
};

ReactFlipContainer.childContextTypes = {
  flip: PropTypes.shape({
    status: PropTypes.func.isRequired,
    registerElement: PropTypes.func.isRequired
  }).isRequired
};

export default ReactFlipContainer;
