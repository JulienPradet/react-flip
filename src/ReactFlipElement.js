import React, { Component } from 'react';
import ReactFlipContainer, { BEFORE_ANIMATION } from './ReactFlipContainer';

const ReactFlipElement = (options = {}) =>
  BaseComponent => {
    const getOptions = props =>
      typeof options === 'function' ? () => options(props) : options;

    class ReactFlipElement extends Component {
      constructor() {
        super();
        this.setFlipElement = this.setFlipElement.bind(this);
        this.updateTarget = this.updateTarget.bind(this);
      }

      componentDidMount() {
        this.removeTarget = this.updateTarget();
      }

      componentDidUpdate() {
        if (
          this.context.flip.status() === BEFORE_ANIMATION &&
          getOptions(this.props).defer
        ) {
          if (this.removeTarget) this.removeTarget();
          this.removeTarget = this.updateTarget();
        }
      }

      componentWillUnmount() {
        if (this.removeTarget) this.removeTarget();
      }

      setFlipElement(element) {
        this.element = element;
      }

      updateTarget() {
        if (!this.element) return;

        return this.context.flip.registerElement({
          element: this.element,
          options: getOptions(this.props)
        });
      }

      render() {
        return (
          <BaseComponent
            flip={{
              setFlipElement: this.setFlipElement,
              status: this.context.flip.status()
            }}
            {...this.props}
          />
        );
      }
    }

    ReactFlipElement.contextTypes = ReactFlipContainer.childContextTypes;

    return ReactFlipElement;
  };

export default ReactFlipElement;
