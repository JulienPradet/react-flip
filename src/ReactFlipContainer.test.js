import React, { Component } from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReactFlipContainer from './ReactFlipContainer';
import ReactFlipElement from './ReactFlipElement';
import FlipGroup from './FlipGroup';

jest.mock('./FlipGroup');

const ReflessElement = ReactFlipElement({})(props => (
  <div>
    Element
  </div>
));

const Element = ReactFlipElement({})(props => (
  <div ref={props.flip.setFlipElement}>
    Element
  </div>
));

class Wrapper extends Component {
  constructor() {
    super();
    this.state = { opened: false };
    this.toggle = this.toggle.bind(this);
  }
  toggle() {
    this.setState(state => ({ opened: !state.opened }));
  }
  render() {
    return (
      <ReactFlipContainer>
        {({ animating }) => (
          <div>
            {animating ? 'Animated' : 'Static'}
            <Element />
          </div>
        )}
      </ReactFlipContainer>
    );
  }
}

describe('ReactFlipContainer', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('The container should render its children', () => {
    const tree = shallow(
      <ReactFlipContainer>
        {({ animating }) => (
          <div>
            {animating ? 'Animated' : 'Static'}
            Content
          </div>
        )}
      </ReactFlipContainer>
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  test('Should allow children to register to flip group', () => {
    const Element = ReactFlipElement({})(props => (
      <div ref={props.flip.setFlipElement}>
        Element
      </div>
    ));

    const tree = mount(
      <ReactFlipContainer>
        {({ animating }) => (
          <div>
            {animating ? 'Animated' : 'Static'}
            <Element />
          </div>
        )}
      </ReactFlipContainer>
    );

    expect(FlipGroup.prototype.addElement.mock.calls.length).toBe(1);
  });

  test('Should not register children if the ref is not defined', () => {
    const tree = mount(
      <ReactFlipContainer>
        {({ animating }) => (
          <div>
            {animating ? 'Animated' : 'Static'}
            <ReflessElement />
          </div>
        )}
      </ReactFlipContainer>
    );

    expect(FlipGroup.prototype.addElement.mock.calls.length).toBe(0);
  });

  test('Should call first, last and invert on update, but not play if invert returned false', () => {
    const tree = mount(<Wrapper />);
    tree.instance().toggle();

    expect({
      first: FlipGroup.prototype.first.mock.calls.length,
      last: FlipGroup.prototype.last.mock.calls.length,
      invert: FlipGroup.prototype.invert.mock.calls.length,
      play: FlipGroup.prototype.play.mock.calls.length
    }).toEqual({
      first: 1,
      last: 1,
      invert: 1,
      play: 0
    });
  });

  test('Should call first, last and invert on update, and play if invert returned true', () => {
    FlipGroup.prototype.invert.mockImplementation(() => true);

    const tree = mount(<Wrapper />);
    tree.instance().toggle();

    expect({
      first: FlipGroup.prototype.first.mock.calls.length,
      last: FlipGroup.prototype.last.mock.calls.length,
      invert: FlipGroup.prototype.invert.mock.calls.length,
      play: FlipGroup.prototype.play.mock.calls.length
    }).toEqual({
      first: 1,
      last: 1,
      invert: 1,
      play: 1
    });
  });

  test('Should render as animating if play returned a promise', () => {
    FlipGroup.prototype.invert.mockImplementation(() => true);
    FlipGroup.prototype.play.mockImplementation(() => new Promise(() => {}));

    const tree = mount(<Wrapper />);
    tree.instance().toggle();

    expect(toJson(tree)).toMatchSnapshot();
  });

  test('Should render back as static if the play promise was resolved', () => {
    let externalResolve;
    FlipGroup.prototype.invert.mockImplementation(() => true);
    FlipGroup.prototype.play.mockImplementation(
      () => new Promise(resolve => externalResolve = resolve)
    );

    const tree = mount(<Wrapper />);
    tree.instance().toggle();
    externalResolve();

    return new Promise(resolve => {
      setTimeout(
        () => {
          expect(toJson(tree)).toMatchSnapshot();
          resolve();
        },
        0
      );
    });
  });

  test('Should remove an element the FlipGroup on unmount', () => {
    const removeMock = jest.fn();
    FlipGroup.prototype.addElement.mockImplementation(() => removeMock);

    class Wrapper extends Component {
      constructor() {
        super();
        this.state = { opened: true };
        this.close = this.close.bind(this);
      }
      close() {
        this.setState({ opened: false });
      }
      render() {
        return (
          <ReactFlipContainer>
            {({ animating }) => (
              <div>
                {this.state.opened && <Element />}
              </div>
            )}
          </ReactFlipContainer>
        );
      }
    }
    const tree = mount(<Wrapper />);
    tree.instance().close();

    expect(removeMock.mock.calls.length).toBe(1);
  });
});
