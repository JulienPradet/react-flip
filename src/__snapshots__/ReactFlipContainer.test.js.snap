// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactFlipContainer Should animate if shouldAnimate is a function that returns true 1`] = `
<Wrapper
  shouldAnimate={[Function]}
>
  <ReactFlipContainer
    defer={false}
    forceDefer={false}
    shouldAnimate={[Function]}
  >
    <div>
      <ReactFlipElement
        opened={true}
      >
        <Component
          flip={
            Object {
              "setFlipElement": [Function],
              "status": "ReactFlip_animation",
            }
          }
          opened={true}
        >
          <div>
            Element
          </div>
        </Component>
      </ReactFlipElement>
    </div>
  </ReactFlipContainer>
</Wrapper>
`;

exports[`ReactFlipContainer Should animate if shouldAnimate is true 1`] = `
<Wrapper
  shouldAnimate={true}
>
  <ReactFlipContainer
    defer={false}
    forceDefer={false}
    shouldAnimate={true}
  >
    <div>
      <ReactFlipElement
        opened={true}
      >
        <Component
          flip={
            Object {
              "setFlipElement": [Function],
              "status": "ReactFlip_animation",
            }
          }
          opened={true}
        >
          <div>
            Element
          </div>
        </Component>
      </ReactFlipElement>
    </div>
  </ReactFlipContainer>
</Wrapper>
`;

exports[`ReactFlipContainer Should not animate if shouldAnimate is a function that returns false 1`] = `
<Wrapper
  shouldAnimate={[Function]}
>
  <ReactFlipContainer
    defer={false}
    forceDefer={false}
    shouldAnimate={[Function]}
  >
    <div>
      <ReactFlipElement
        opened={true}
      >
        <Component
          flip={
            Object {
              "setFlipElement": [Function],
              "status": "ReactFlip_static",
            }
          }
          opened={true}
        >
          <div>
            Element
          </div>
        </Component>
      </ReactFlipElement>
    </div>
  </ReactFlipContainer>
</Wrapper>
`;

exports[`ReactFlipContainer Should not animate if shouldAnimate is false 1`] = `
<Wrapper
  shouldAnimate={false}
>
  <ReactFlipContainer
    defer={false}
    forceDefer={false}
    shouldAnimate={false}
  >
    <div>
      <ReactFlipElement
        opened={true}
      >
        <Component
          flip={
            Object {
              "setFlipElement": [Function],
              "status": "ReactFlip_static",
            }
          }
          opened={true}
        >
          <div>
            Element
          </div>
        </Component>
      </ReactFlipElement>
    </div>
  </ReactFlipContainer>
</Wrapper>
`;

exports[`ReactFlipContainer Should render as animating if play returned a promise 1`] = `
<Wrapper>
  <ReactFlipContainer
    defer={false}
    forceDefer={false}
  >
    <div>
      <ReactFlipElement
        opened={true}
      >
        <Component
          flip={
            Object {
              "setFlipElement": [Function],
              "status": "ReactFlip_animation",
            }
          }
          opened={true}
        >
          <div>
            Element
          </div>
        </Component>
      </ReactFlipElement>
    </div>
  </ReactFlipContainer>
</Wrapper>
`;

exports[`ReactFlipContainer Should render back as static if the play promise was resolved 1`] = `
<Wrapper>
  <ReactFlipContainer
    defer={false}
    forceDefer={false}
  >
    <div>
      <ReactFlipElement
        opened={true}
      >
        <Component
          flip={
            Object {
              "setFlipElement": [Function],
              "status": "ReactFlip_static",
            }
          }
          opened={true}
        >
          <div>
            Element
          </div>
        </Component>
      </ReactFlipElement>
    </div>
  </ReactFlipContainer>
</Wrapper>
`;

exports[`ReactFlipContainer The container should render its children 1`] = `
<div>
  Content
</div>
`;
